### PR TITLE
FEATURE: Add user status to inline mentions in chat

### PR DESCRIPTION
**Pull request by AndrewPrigorshnev**
Screenshot 2023-05-01 at 22 38 26

This PR adds status to mentions in chat and makes those mentions receive live updates.

There are known unfinished part in this implementation: when posting a message, status on mentions on that message appears immediately, but only if a user used autocomplete when typing the message. If user copy and paste a message with mentions into chat composer, those mentions won't have user status on them.

PRs with fixes for both problems are following soon.

Preparations for this PR that were made previously include:
- DEV: correct a relationship – a chat message may have several mentions 0dcfd7ddeccd438fed97c15827214a3ddd944838
- DEV: extract the logic for extracting and expanding mentions from ChatNotifier 75b81b6854087842b53b4c9559ef5836d9516269
- DEV: Always create chat mention records fa543cda06594b5bebe0ab35e48e613540f9d3dc
- DEV: better split create_notification! and send_notifications logic e292c45924bb0b0a79497e5f494afcb8af2e1efc
- DEV: more tests for mentions when updating chat messages e7292e1682bb910635af26879c31a17b8843b738
- DEV: extract updating status on mentions into a lib function e49d338c21885189970e51a60ac79ab8c1fc397e
- DEV: Create and update chat message mentions earlier 35a414bb38b2ab990ff920f8c3adf2f877249f12
- DEV: Create a chat_mention record when self mentioning 2703f2311aefd87a7774316b50f5339626ea5b48
- DEV: When deleting a chat message, do not delete mention records f4fde4e49b03d5e8f01d02e2354c553c82f3402d

### CODE REVIEW

**Comment by jjaffeux** [comment]
@AndrewPrigorshnev I have fixed conflicts for you we should try to move on this when you can

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js [hunk 1]
```diff
@@ -86,6 +96,25 @@ export default class ChatMessage extends Component {
   @action
   teardownChatMessage() {
     cancel(this._invitationSentTimer);
+
+    this.args.message.mentionedUsers.forEach((user) => {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.js:100]: Can we abstract all of this in a method of message? teardownMentionedUsers ?

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js [hunk 2]
```diff
@@ -86,6 +96,25 @@ export default class ChatMessage extends Component {
...
+      user.stopTrackingStatus();
+      user.off("status-changed", this, "refreshStatusOnMentions");
+    });
+  }
+
+  @action
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.js:106]: can we use the decorators hooks for this? each time a message is updated the version will be incremented and decorates hooks will be replayed, look at what we do for dates for example

File: plugins/chat/assets/javascripts/discourse/routes/chat.js [hunk 3]
```diff
@@ -65,8 +65,10 @@ export default class ChatRoute extends DiscourseRoute {
   }
 
   deactivate(transition) {
-    const url = this.router.urlFor(transition.from.name);
-    this.chatStateManager.storeChatURL(url);
+    if (transition) {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/routes/chat.js:68]: can you expand on when does this case happen please?

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js (hunk 1 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.js]: That will be good, done.

File: plugins/chat/assets/javascripts/discourse/routes/chat.js (hunk 3 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/assets/javascripts/discourse/routes/chat.js:68]: It happens when `deactivate` is called from the route's `willDestroy` method, in that case, there is no transition. So the transition argument is nullable:
https://github.com/emberjs/ember.js/blob/06735dbba898457afc9acbc551ae389d06b02306/packages/%40ember/routing/route.ts#L857

This place fails in acceptance tests, because at the end of an acceptance test, Ember just destroys the app and the route.

As far as I know this will never happen in a real app, for example see https://discuss.emberjs.com/t/are-routes-destroyed-while-transitioning/17644.

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js (hunk 2 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.js:98]: This is tricky. I initially implemented it as a decorator in `initializers/chat-decorators.js`, then I felt that it's better to reimplement it this way.

Note that we need to re-render status in the next cases:
1. when a message is posted
2. when a message is edited
3. when we receive a message bus message with status update for a mentioned user

When I implemented it as a decorator, I needed to redecorate the message in the third case, which calls all decorators, while in fact there is no need to re-render everything, it's enough to only update user status.

By the way, as a side note, another thing that bothers me is that these decorators are implemented using an initializer. That couples the chat-message component with the application startup logic. You can notice that if you try to write a component test for the `chat-message` component, none of the decorators will work by default. Of course, we can initialize decorators before running a component test, but in general, the fact that we can't just test a component independently is a sign that that logic actually naturally belongs to the component itself. I think it's something we can think about, looks like it's possible to reimplement these decorators without using initializer, that may be simpler.

So maybe let's better to leave re-rendering using status in the component as it's implemented now? It's more efficient, easier to understand and easier to test?

**Comment by jjaffeux** [comment]
Correct me if Im wrong but it appears that if change my status while looking at a channel containing a mention of my username it will show the status on it, but will disappear after reload and never appear again? Is that expected?

**Comment by AndrewPrigorshnev** [comment]
I've found another problem – messages comes from message bus without status. Status shows correctly in these cases:
1. When you load a channel that already contains messages
2. When you're in a channel and type a message that's mentioning someone (status will appear immediately)

But if you look at a channel and at the same time someone's typing a message that's mentioning you, you'll receive that message without status on it.

That shouldn't be hard to fix, I'm going to fix it in this PR tomorrow.

**Comment by AndrewPrigorshnev** [comment]
> But if you're looking at a channel and at the same time someone's typing a message that's mentioning you, you'll receive that message without status on it.
> 
> That shouldn't be hard to fix, I'm going to fix it in this PR tomorrow.

This needs some more refactoring. When I was decoupling `chat_mentions` from `notifications`, I was trying to be careful (maybe too careful :)), so I didn't move mentions creation from `chat_notfier`. Now we need to create them earlier – before we send a message back to clients via MessageBus.

I did that in https://github.com/discourse/discourse/pull/21388, let's review and merge that PR first.

**Comment by AndrewPrigorshnev** [comment]
@jjaffeux I've pushed commits that handle various edge cases and add tests for them. I also handled status on edited messages (initially I planned to fix it later in a separate PR).

File: plugins/chat/assets/javascripts/discourse/components/chat-message.js [hunk 4]
```diff
@@ -104,6 +110,9 @@ export default class ChatMessage extends Component {
     };
 
     this.args.message.expanded = true;
+    schedule("afterRender", () => {
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/chat-message.js:113]: I would most likely have the after render directly in `refreshStatusOnMentions` function, you shouldn't need to have the knowledge that this function needs to be wrapped in an after render block

Comment (AndrewPrigorshnev) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.js:113]: I think, more precisely, this `schedule("afterRender"` isn't related to `refreshStatusOnMentions`, it's related to the line above:

`this.args.message.expanded = true;`

At this moment, the message is collapsed, we're expanding it by setting `expanded` to `true`. If we want to do something with the expanded message in UI right away (it can be a call to `refreshStatusOnMentions` or it can be something else), we have to schedule that to the afterRender queue.

Other case when we call `refreshStatusOnMentions` is from modifiers, and in those cases there is no need to schedule it to the after render queue.

I feel, from this point of view, this is a good place for this call?

Comment (jjaffeux) [reply plugins/chat/assets/javascripts/discourse/components/chat-message.js:113]: I see it the opposite way, the code in `refreshStatusOnMentions` needs to act on the DOM, as a result we need to always be sure we are not in the middle of changing the DOM. This is always going to be related to something outside of `refreshStatusOnMentions`, but leaving the responsibility to the developer to call schedule before calling it means you need to know the implementation of `refreshStatusOnMentions` and even details of it, before calling it.

File: app/assets/javascripts/discourse/app/models/user.js [hunk 5]
```diff
@@ -1436,6 +1436,10 @@ User.reopen(Evented, {
     this._subscribersCount--;
   },
 
+  isTrackingStatus() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/models/user.js:1439]: would be very nice at some point to have all of this using glimmer and tracked properties, but too much scope creep for this PR

File: plugins/chat/app/controllers/chat/chat_controller.rb [hunk 6]
```diff
@@ -412,6 +412,7 @@ def preloaded_chat_message_query
           .includes(:uploads)
           .includes(chat_channel: :chatable)
           .includes(:thread)
+          .includes(:chat_mentions)
```
Comment (jjaffeux) [inline plugins/chat/app/controllers/chat/chat_controller.rb:415]: You probably need to add it chat messages query following recent @martin changes

**Review by jjaffeux — APPROVED** [review]

**Comment by jjaffeux** [comment]
Frontend tests failures look legit

File: plugins/chat/app/controllers/chat/chat_controller.rb (hunk 6 above)
Comment (AndrewPrigorshnev) [reply plugins/chat/app/controllers/chat/chat_controller.rb:415]: It would be nice if in such cases we refactor logic in place instead of creating a full duplicate of the logic. I'll add this to the new query, but I'm going to need to also make sure that everything is works as expected and properly tested.

Comment (jjaffeux) [reply plugins/chat/app/controllers/chat/chat_controller.rb:415]: We just can't all remove at once

Comment (jjaffeux) [reply plugins/chat/app/controllers/chat/chat_controller.rb:415]: Sure feel free to do so, we will remove this controller endpoint soon, but no problem from me if you want to do this :+1:

File: app/assets/javascripts/discourse/app/models/user.js (hunk 5 above)
Comment (AndrewPrigorshnev) [reply app/assets/javascripts/discourse/app/models/user.js:1439]: Yes, if we have glimmer components everywhere, handling this will be simpler.
